x/programs/rust/examples/nft/src/lib.rs

```diff
@@ -0,0 +1,158 @@
+    public, state_keys,
+    types::Address,
+};
+
+/// The max supply of the NFT.
+const MAX_SUPPLY: i64 = 10;
+
+/// The program storage keys.
+#[state_keys]
+enum StateKey {
+    /// The total supply of the token. Key prefix 0x0.
+    MaxSupply,
+    /// The name of the token. Key prefix 0x1.
+    Name,
+    /// The symbol of the token. Key prefix 0x2.
+    Symbol,
+    /// Metadata of the token. Key prefix 0x3.
+    Uri,
+    /// Balances of the NFT token by address. Key prefix 0x4(address).
+    Balances(Address),
+    /// Counter -- used to keep track of total NFTs minted. Key prefix 0x5.
+    Counter,
+}
+
+/// Initializes the NFT with all required metadata.
+/// This includes the name, symbol, image URI, owner, and total supply.
+/// Returns true if the initialization was successful.
+#[public]
+pub fn init(
+    program: Program,
+    nft_name_ptr: i64,
+    nft_name_length: i64,
+    nft_symbol_ptr: i64,
+    nft_symbol_length: i64,
+    nft_uri_ptr: i64,
+    nft_uri_length: i64,
+) -> bool {
+    let name_ptr = Memory::new(Pointer::from(nft_name_ptr));
+    let nft_name = unsafe { name_ptr.range(nft_name_length as usize) };
+
+    let nft_symbol_ptr = Memory::new(Pointer::from(nft_symbol_ptr));
+    let nft_symbol = unsafe { nft_symbol_ptr.range(nft_symbol_length as usize) };
+
+    let nft_uri_ptr = Memory::new(Pointer::from(nft_uri_ptr));
+    let nft_uri = unsafe { nft_uri_ptr.range(nft_uri_length as usize) };
+
+    let counter = program
+        .state()
+        .get::<i64, _>(StateKey::Counter.to_vec())
+        .unwrap_or_default();
+
+    assert_eq!(counter, 0, "init already called");
+
+    // Set token name
+    program
+        .state()
+        .store(StateKey::Name.to_vec(), &nft_name)
+        .expect("failed to store nft name");
+
+    // Set token symbol
+    program
+        .state()
+        .store(StateKey::Symbol.to_vec(), &nft_symbol)
+        .expect("failed to store nft symbol");
+
+    // Set token URI
+    program
+        .state()
+        .store(StateKey::Uri.to_vec(), &nft_uri)
+        .expect("failed to store nft uri");
+
+    // Set total supply
+    program
+        .state()
+        .store(StateKey::MaxSupply.to_vec(), &MAX_SUPPLY)
+        .expect("failed to store total supply");
+
+    // Initialize counter
+    program
+        .state()
+        .store(StateKey::Counter.to_vec(), &0_i64)
+        .expect("failed to store counter");
+
+    true
+}
+
+/// Mints NFT tokens and sends them to the recipient.
+#[public]
+pub fn mint(program: Program, recipient: Address, amount: i64) -> bool {
+    let counter = program
+        .state()
+        .get::<i64, _>(StateKey::Counter.to_vec())
+        .unwrap_or_default();
+
+    assert!(counter <= MAX_SUPPLY, "max supply for nft exceeded");
+
+    let balance = program
+        .state()
+        .get::<i64, _>(StateKey::Balances(recipient).to_vec())
+        .unwrap_or_default();
+
+    assert!(balance + amount <= MAX_SUPPLY, "mint exceeded max supply");
```

Comment:
this check is not necessary (`MAX_SUPPLY` relates to the total count, not an individual user's balance)